shadowsocks/udp.go

```diff
@@ -187,45 +188,60 @@ func (s *udpService) Stop() error {
 }
 
 type natentry struct {
-	conn net.PacketConn
+	Conn net.PacketConn
 	// We store the client location in the NAT map to avoid recomputing it
 	// for every outbound packet in a UDP-based connection.
 	clientLocation string
+	// Number of upstream packets sent on Conn.
+	count int64
+	// Number of DNS query packets sent on Conn.
+	countDNS int64
 }
 
 // Packet NAT table
 type natmap struct {
 	sync.RWMutex
-	keyConn map[string]natentry
+	keyConn map[string]*natentry
 	timeout time.Duration
 	metrics metrics.ShadowsocksMetrics
 }
 
 func newNATmap(timeout time.Duration, sm metrics.ShadowsocksMetrics) *natmap {
 	m := &natmap{metrics: sm}
-	m.keyConn = make(map[string]natentry)
+	m.keyConn = make(map[string]*natentry)
 	m.timeout = timeout
 	return m
 }
 
-func (m *natmap) Get(key string) (net.PacketConn, string) {
+func (m *natmap) Get(key string) *natentry {
 	m.RLock()
 	defer m.RUnlock()
-	entry := m.keyConn[key]
-	return entry.conn, entry.clientLocation
+	return m.keyConn[key]
 }
 
 // Refresh the NAT mapping.  This should be called on every write for
 // outbound-refresh behavior (as required by RFC 4787 Section 4.3).
-func (m *natmap) Refresh(targetConn net.PacketConn) {
-	targetConn.SetReadDeadline(time.Now().Add(m.timeout))
+func (m *natmap) Refresh(entry *natentry, dstPort int) {
```

Comment:
Let's have a PacketConn Wrapper with WriteTo method that takes care of the NAT refresh.
It should have ReadFrom() too, and maybe an expired or closed method, so you can pass to timedCopy.

This way timedCopy doesn't need to know about refresh logic and optimizations.

Reply:
Such wrapper can hold the location and take care of metrics too, effectively removing the need for natentry.

Reply:
Maybe we have a different wrapper implementation for DNS, and we pick that implementation when we create the wrapper, based on the port number. This way the DNS optimization is encapsulated.

Reply:
This is elegant but it's not very compatible with mixed DNS and non-DNS connections, as we discussed.  It seems like we decided to try to support that case, so I didn't implement this pattern.